File: Compiler design/Lab4/Java/src/ru/bmstu/iu9/Scanner.java
package ru.bmstu.iu9;


public class Scanner {
    public final String program;

    private Compiler compiler;
    private Position cur;

    public Scanner(String program, Compiler compiler) {
        this.compiler = compiler;
        this.program = program;
        cur = new Position(program);
    }

    public Token nextToken() {
        while (!cur.isEOF()) {
            while (cur.isWhitespace())
                cur = cur.next();
            Token token;
            switch (cur.getCode()) {
                case ':':
                    if (cur.next().getCode() == ':')
                        token = readComment(cur);
                    else
                        token = readLabel(cur);
                    break;
                case '\"':
                    token = readWord(cur,true);
                    break;
                default:
                    token = readWord(cur,false);
                    break;
            }
            if (token == null || token.tag == DomainTag.UNKNOWN) {
                compiler.addMessage(true,cur,"Token: unrecognized token");
                cur = cur.next();
            }
            else {
                cur = token.coords.following;
                return token;
            }
        }
        return new UnknownToken(DomainTag.END_OF_PROGRAM,cur,cur);
    }

    private Token readComment(Position cur) {
        Position p = cur.next().next();
        StringBuilder sb = new StringBuilder();
        while (true) {
            if (p.isNewLine())
                break;
            else {
                sb.append(Character.toChars(p.getCode()));
                p = p.next();
            }
        }
        return new CommentToken(sb.toString(),cur,p.next());
    }

    private Token readLabel(Position cur) {
        StringBuilder sb = new StringBuilder();
        Position p = cur.next();
        if (p.isLetter() || p.isDecimalDigit()) {
            while (p.isLetter() || p.isDecimalDigit()) {
                sb.append(Character.toChars(p.getCode()));
                p = p.next();
            }
            return new LabelToken(sb.toString(), cur, p.next());
        }
        compiler.addMessage(false,cur,"Label must start only with numbers or letters");
        return new LabelToken(sb.toString(),cur,p.next());
    }

    private Token readWord(Position cur,boolean isQuote) {
        if (!isQuote) {
            Position p = cur.next();
            StringBuilder sb = new StringBuilder();
            sb.append(Character.toChars(cur.getCode()));
            while (!p.isEOF() && !p.isWhitespace()) {
                sb.append(Character.toChars(p.getCode()));
                p = p.next();
            }
            return new WordToken(sb.toString(),cur,p);
        }
        else {
            Position p = cur.next();
            StringBuilder sb = new StringBuilder();
            while (!p.isEOF() && !(p.getCode()=='\"')) {
                if (p.getCode() == '\r' || p.getCode() == '\n')
                    break;
                else {
                    sb.append(Character.toChars(p.getCode()));
                    p = p.next();
                }
            }
            if (p.getCode() == '\"')
                return new WordToken(sb.toString(),cur,p.next());
            else if (p.getCode() == '\r') {
                compiler.addMessage(false,p,"words don't contains whitespaces, \" expected");
                return new WordToken(sb.toString(),cur,p);
            }
            else {
                compiler.addMessage(false,p,"end of program found, \" expected");
                return new WordToken(sb.toString(),cur,p);
            }
        }
    }
}

File: Compiler design/Lab4/OCaml/Scanner.ml
open Batteries;;
open DomainTag;;


class scanner programm compileer = object (self)
    
    val program: BatUTF8.t = programm
    val compiler: Compiler.compiler = compileer
    val mutable cur: Position.position = new Position.position (Position.init_pos programm)

    method cur = cur
  

    method read_comment (pos : Position.position) =
        let new_pos = ref (Position.create_pos pos#get_text pos#get_line pos#get_pos pos#get_index) in
        let p = ref (new Position.position !new_pos)#next#next in
        let s = ref "" in
        while (not !p#is_newline) do
            s := !s ^ BatChar.escaped (BatChar.chr !p#get_code);
            p := !p#next;
        done;
        new CommentToken.token cur !p#next !s

    method read_label (pos : Position.position) =
        let new_pos = ref (Position.create_pos pos#get_text pos#get_line pos#get_pos pos#get_index) in
        let p = ref (new Position.position !new_pos)#next in
        let s = ref "" in
        if (!p#is_decimaldigit || !p#is_letter) then (
            while (!p#is_decimaldigit || !p#is_letter) do
                s := !s ^ BatChar.escaped (BatChar.chr !p#get_code);
                p := !p#next;
            done;
            new LabelToken.token cur !p#next !s;
        )
        else (
            compiler#add_message false !p "Label must start only with numbers or letters\n";
            new UnknownToken.token Unknown cur !p#next !s;
        );

    method read_word (pos : Position.position) (flag: bool) =
        if (not flag) then (
            let new_pos = ref (Position.create_pos pos#get_text pos#get_line pos#get_pos pos#get_index) in
            let p = ref (new Position.position !new_pos)#next in
            let s = ref "" in
            s := !s ^ BatChar.escaped (BatChar.chr pos#get_code);
            while (not !p#is_eof && not !p#is_whitespace) do
                s := !s ^ BatChar.escaped (BatChar.chr !p#get_code);
                p := !p#next;
            done;
            new WordToken.token cur !p !s
        )
        else (
            let new_pos = ref (Position.create_pos pos#get_text pos#get_line pos#get_pos pos#get_index) in
            let p = ref (new Position.position !new_pos)#next in
            let s = ref "" in
            while (not !p#is_eof && (BatChar.chr !p#get_code != '\"') && (BatChar.chr !p#get_code != '\r') && (BatChar.chr !p#get_code != '\n')) do
                s := !s ^ BatChar.escaped (BatChar.chr !p#get_code);
                p := !p#next;
            done;
            if (BatChar.chr !p#get_code == '\"') then
                new WordToken.token cur !p#next !s
            else if (!p#is_eof) then begin
                compiler#add_message false !p "end of program found, \" expected\n";
                new UnknownToken.token Unknown cur !p !s
            end
            else if ((BatChar.chr !p#get_code == '\r') || (BatChar.chr !p#get_code == '\n')) then begin
                compiler#add_message false !p "Words don't contains whitespaces, \" expected\n";
                new UnknownToken.token Unknown cur !p !s
            end
            (* this isn't necessary, but compiler uses automatic type deduction, so you must have 'else' - statement *)
            else 
                new UnknownToken.token Unknown cur !p !s
        );
   
    method next_token =
        let tok = ref (new UnknownToken.token Unknown cur cur "") in
        while !tok#tag == Unknown do
            while cur#is_whitespace do
                cur <- cur#next
            done;
            if cur#is_eof then
                tok := (new UnknownToken.token End_of_program cur cur "")
            else begin
                let token =
                    match BatChar.chr cur#get_code with
                        | ':' when cur#next#get_code == BatChar.code(':') -> self#read_comment cur 
                        | ':'                                             -> self#read_label cur
                        | '\"'                                            -> self#read_word cur true
                        | _                                               -> self#read_word cur false
                    in
                    if (token#tag == Unknown) then begin
                        compiler#add_message true cur "Token: unrecognized token\n";
                        cur <- cur#next;
                    end
                    else begin
                        cur <- token#coords#following;
                        tok := (token :> Token.token);               
                    end;
            end
        done;
        !tok
    end;;
